sdks/java/core/src/main/java/org/apache/beam/sdk/io/DefaultFilenamePolicy.java

```diff
@@ -182,19 +184,26 @@ public void encode(Params value, OutputStream outStream) throws IOException {
       if (value == null) {
         throw new CoderException("cannot encode a null value");
       }
-      stringCoder.encode(value.baseFilename.get().toString(), outStream);
-      stringCoder.encode(value.shardTemplate, outStream);
-      stringCoder.encode(value.suffix, outStream);
+      STRING_CODER.encode(value.baseFilename.get().toString(), outStream);
+      STRING_CODER.encode(value.shardTemplate, outStream);
+      STRING_CODER.encode(value.suffix, outStream);
+      BOOLEAN_CODER.encode(value.baseFilename.get().isDirectory(), outStream);
     }
 
     @Override
     public Params decode(InputStream inStream) throws IOException {
-      ResourceId prefix =
-          FileBasedSink.convertToFileResourceIfPossible(stringCoder.decode(inStream));
-      String shardTemplate = stringCoder.decode(inStream);
-      String suffix = stringCoder.decode(inStream);
+      String prefix = STRING_CODER.decode(inStream);
+      String shardTemplate = STRING_CODER.decode(inStream);
+      String suffix = STRING_CODER.decode(inStream);
+      ResourceId baseFilename;
+      if (inStream.available() > 0) {
+        baseFilename = FileSystems.matchNewResource(prefix, BOOLEAN_CODER.decode(inStream));
+      } else {
+        // fallback for ensure backward compatibility
+        baseFilename = FileBasedSink.convertToFileResourceIfPossible(prefix);
```

Comment:
Would it be possible to add test case for bw compatible code path?

Reply:
ok, I added the requested test for bw compatibility

btw you can see the buggy behavior there (lost information whether baseFilename is file/directory)

Reply:
convertToFileResourceIfPossible attempts to match a file and if that fails attempts to match a directory.

Is the issue that the underlying filesystem erroneously says something is a file when really it is a directory?

Reply:
Exactly, that's the issue, for example `LocalFileSystem` matches path `/tmp/dirA/` as file..

How about something like this?
```java
public static class ParamsCoder extends AtomicCoder<Params> {
    private static final ParamsCoder INSTANCE = new ParamsCoder();
    private static final Coder<String> STRING_CODER = StringUtf8Coder.of();

    public static ParamsCoder of() {
      return INSTANCE;
    }

    @Override
    public void encode(Params value, OutputStream outStream) throws IOException {
      if (value == null) {
        throw new CoderException("cannot encode a null value");
      }
      ResourceId baseFilename = value.getBaseFilename().get();
      String baseFilenameString =
          (baseFilename.isDirectory() && !baseFilename.toString().endsWith("/"))
              ? baseFilename.toString() + "/"
              : baseFilename.toString();
      STRING_CODER.encode(baseFilenameString, outStream);
      STRING_CODER.encode(value.getShardTemplate(), outStream);
      STRING_CODER.encode(value.getSuffix(), outStream);
    }

    @Override
    public Params decode(InputStream inStream) throws IOException {
      String baseFilenameString = STRING_CODER.decode(inStream);
      ResourceId prefix =
          FileSystems.matchNewResource(baseFilenameString, baseFilenameString.endsWith("/"));
      String shardTemplate = STRING_CODER.decode(inStream);
      String suffix = STRING_CODER.decode(inStream);
      return new Params()
          .withBaseFilename(prefix)
          .withShardTemplate(shardTemplate)
          .withSuffix(suffix);
    }
  }
```
But I'm not sure how to handle the suffix (`/`) to ensure compatibility among (file) systems

Reply:
I think that there are a couple of options here:
1) Take the breaking change to the coder because it is a fix, make sure it is documented in the release notes
2) Try to fix the underlying filesystem to do a better job of file/dir matching
3) Deprecate this filename policy, create a new one (DefaultFilenamePolicy2) and tell people to use it in new code.

I'm for #1 but would ask for consensus on the mailing list.

Reply:
OK, I agree with the first solution - taking the breaking change. I will ask on the mailing list.

BTW I did workaround in my project where I use only files (not directories) as `Params` base filenames and then everything works fine. That means I no longer needs this fix but it still seems to me like a bug which is worth fixing.

(P.S.: The suffix wouldn't have to be the path separator (`/`) but e.g. `sometotallyrandomstring123`  - but it's very ugly...)

Reply:
https://lists.apache.org/thread.html/r1d844830648b88948d38d0c28d106dd7fe2ed92af4a170e2679f8c2f%40%3Cdev.beam.apache.org%3E

Reply:
Hi @lukecwik, finally after the discussion in the mailing list I chose the second proposed option and fixed the underlying filesystem. Turns out that only `LocalFileSystem` is broken, on the other hand e.g. HDFS and S3 look fine and already have this kind of check.

I still think that the proper solution should be to use the `ResourceIdCoder` instead of `StringUtf8Coder` but I understand the consequences.